### PR TITLE
FileManager+unzip: Unzip from context menu

### PR DESCRIPTION
**Pull request by TheGeopard**
Can't really call this my first PR anymore after forgetting to format the old one properly :facepalm:, sorry for the trouble, still new to this.

This PR adds a new context menu action to FileManager that allows for extracting zip archives without having to open the terminal. I thought this would make a good first contribution after playing around a bit with SerenityOS and finding this quality of life feature that can be found in other file managers like those of Gnome and Windows to be lacking.

"Extract Here" in Gnome
Screenshot from 2021-04-24 13-40-33

My implementation in SerenityOS
Screenshot from 2021-04-25 14-30-27

Future improvements may include support for more archive formats like .tar and .tar.gz as well as integration with the new file operation progress tracking.

### CODE REVIEW

**Comment by IdanHo** [comment]
nice work, but for next time: instead of opening a new PR, you can force push to the branch your PR tracks to replace the commits in the PR :)

File: Userland/Applications/FileManager/main.cpp [hunk 1]
```diff
@@ -192,6 +194,32 @@ void do_create_link(const Vector<String>& selected_file_paths, GUI::Window* wind
...
+        if (rc < 0) {
+            perror("execlp");
+            _exit(1);
+        }
+    } else {
+        // FIXME: this could probably be tied in with the new file operation progress tracking
+        int status;
+        waitpid(unzip_pid, &status, 0);
+        if (status != 0)
+            GUI::MessageBox::show(window, "Could not extract archive", "Extract Archive Error", GUI::MessageBox::Type::Error);
```
Comment (linusg) [inline Userland/Applications/FileManager/main.cpp:219]: `waitpid()` itself needs an error check as well, also I think we should use `WIFEXITED` and `WEXITSTATUS` here:

```suggestion
        int status;
        int rc = waitpid(unzip_pid, &status, 0);
        if (rc < 0 || !WIFEXITED(status) || WEXITSTATUS(status) != 0)
            GUI::MessageBox::show(window, "Could not extract archive", "Extract Archive Error", GUI::MessageBox::Type::Error);
```

File: Userland/Applications/FileManager/main.cpp [hunk 2]
```diff
@@ -354,6 +394,11 @@ int run_in_desktop_mode([[maybe_unused]] RefPtr<Core::ConfigFile> config)
                 file_context_menu->add_action(directory_view.delete_action());
                 file_context_menu->add_separator();
 
+                if (node.full_path().ends_with(".zip")) {
```
Comment (linusg) [inline Userland/Applications/FileManager/main.cpp:397]: Let's ensure this also works with `ARCHIVE.ZIP`, i.e. let's do a case-insensitive compare.

```suggestion
                if (node.full_path().ends_with(".zip", CaseSensitivity::CaseInsensitive)) {
```

File: Userland/Applications/FileManager/main.cpp [hunk 3]
```diff
@@ -965,6 +1023,11 @@ int run_in_windowed_mode(RefPtr<Core::ConfigFile> config, String initial_locatio
                 file_context_menu->add_action(shortcut_action);
                 file_context_menu->add_separator();
 
+                if (node.full_path().ends_with(".zip")) {
```
Comment (linusg) [inline Userland/Applications/FileManager/main.cpp:1026]: Same as above

File: Userland/Utilities/unzip.cpp [hunk 4]
```diff
@@ -111,6 +114,20 @@ int main(int argc, char** argv)
         return 1;
     }
 
+    if (!output_directory_path.is_empty()) {
```
Comment (linusg) [inline Userland/Utilities/unzip.cpp:117]: Let's ensure we don't allow `-o ""`, just use the string's null state instead meaning the argument wasn't passed:

```suggestion
    if (!output_directory_path.is_null()) {
```

**Review by linusg — CHANGES_REQUESTED** [review]

**Comment by TheGeopard** [comment]
I have implemented all the suggested changes 👍

**Pull request merged**